### PR TITLE
[leaf_function][print] Add register_multi_grad_hook for leaf function and debug print backwards and multi-tensors

### PR DESCRIPTION
**Pull request by yushangdi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #175453

## Summary

Add a `register_hook` API to `leaf_function` that allows registering backward hooks on leaf functions. This enables side-effect-only leaf functions (e.g., debug logging) to also execute during the backward pass without requiring the caller to capture the return value.

The hook function has the same signature as the leaf function, but each tensor argument receives the corresponding gradient instead of the original tensor. Non-tensor arguments are passed through unchanged. The hook is itself invoked as a leaf function, so it remains opaque to the compiler. Hooks are registered on the outer (pre-detach) tensors in `InvokeLeafFunctionAutogradOp.forward`, so they fire when autograd computes gradients in the enclosing graph. This works across eager, `torch.compile`, and `aot_function` paths.

As an example use case, `torch.utils.debug_log` provides `debug_log` and `debug_log_rank` utilities that print tensor norms during forward and backward, useful for debugging training numerics. `debug_log` uses `torch._higher_order_ops.print` for stdout output; `debug_log_rank` uses Python logging with optional rank filtering for distributed training.

Start by reading `torch/_dynamo/decorators.py` (the `register_hook` API and docstring), then `torch/_higher_order_ops/invoke_leaf_function.py` (the hook registration in autograd), then `torch/utils/debug_log.py` (the example usage).

```
Example::

# user can register their own custom log function the same way

    from torch._dynamo.decorators import leaf_function


    @leaf_function
    def my_log(t, label):

        print(f"[{label}][fwd] mean={t.mean().item():.4f}")
        return None


    @my_log.register_fake
    def my_log_fake(t, label):
        return None


    @my_log.register_hook
    def my_log_hook(t_grad, label):
        print(f"[{label}][bwd] mean={t_grad.mean().item():.4f}")


    # Usage (works in eager, torch.compile with aot_eager backend, and make_fx):
    y = x * 2
    my_log(y, "after_linear")  # logs fwd on call, bwd when y's grad is computed

```

## Test plan

- `python -m pytest test/higher_order_ops/test_debug_log.py -xvs` (22 passed, 1 xfailed)
- Tests cover eager, `torch.compile` (aot_eager), and `aot_function` paths for both `debug_log` and `debug_log_rank`
- Tests verify forward/backward logs appear, FX graphs contain `invoke_leaf_function` nodes, gradients are correct, and rank filtering works
- Expected failure test documents that DTensor cannot be passed directly (HOP has no DTensor dispatch rule)

Authored with Claude.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @mlazos

### CODE REVIEW

File: test/functorch/test_leaf_function.py [hunk 1]
```diff
@@ -766,5 +766,309 @@ def forward(self, arg0_1, arg1_1: "f32[3, 3]"):
...
+
+        x = torch.randn(3, requires_grad=True)
+        y = torch.randn(3, requires_grad=True)
+        out = my_fn(x, y)[0]
+        out.sum().backward()
+
+        # Hook fires once per requires_grad tensor input.
+        # Each call replaces one tensor with its gradient while keeping
+        # the other as the original tensor value.
+        self.assertEqual(len(hook_calls), 2)
```
Comment (ydwu4) [inline test/functorch/test_leaf_function.py]: This seems weird: it would probably be a cleaner interface if the hook is per requires_grad tensor and trigers exactly once?

In that case, how does it compare with users calling t.register_hook themselves at the place where it calls the leaf_function?

**Review by ydwu4 — COMMENTED** [review]
Changes look Ok to me. Have some concerns for the leaf_fn.register_hook api.

File: torch/_higher_order_ops/invoke_leaf_function.py [hunk 2]
```diff
@@ -714,36 +713,51 @@ def fake_backward(*grads):
...
+                                    h_fake,
+                                    in_spec,
+                                    "",
+                                    *new_flat_args,
+                                )
 
                         return hook
 
-                    arg.register_hook(
+                    flat_args[i].register_hook(
```
Comment (ydwu4) [inline torch/_higher_order_ops/invoke_leaf_function.py:753]: there is an register_multi_grad_hook, can try this?

Comment (fxdawnn) [reply torch/_higher_order_ops/invoke_leaf_function.py:753]: ``register_hook`` is a simpler mechanism — it just appends a callback to the tensor's gradient accumulator. When any backward pass computes a gradient for that tensor, the callback fires. It doesn't need to query the autograd engine about which nodes will execute.                           
   
``register_multi_grad_hook`` is more sophisticated — it needs to figure out which tensors will actually receive gradients in this specific backward call (so it can skip tensors that won't participate). To do that, it calls _will_engine_execute_node, which inspects the engine's execution plan. But during AOT autograd's autograd.grad() tracing, that introspection isn't available, so it crashes.

I tried and it turned out to be way more complicated.

Comment (ydwu4) [reply torch/_higher_order_ops/invoke_leaf_function.py:753]: A better front API for users is something like this:
```
        @my_fn.register_multi_grad_hook
        def my_fn_hook(x_or_grad, y_val):
            hook_calls.append((x_or_grad.clone(), y_val.clone()))
```

This is better because this clearly indicates that when all the gradients are ready, the hook is triggered exactly once. The register_hook is kind of confusing because it's not immediate when and how many times the hooks will be triggered.

Implementation wise, register_multi_grad_hook is also the better choice so we don't have to track how many grads are ready.

Comment (fxdawnn) [reply torch/_higher_order_ops/invoke_leaf_function.py:753]: I see! Thank you for this example. I will update.

Comment (fxdawnn) [reply torch/_higher_order_ops/invoke_leaf_function.py:753]: updated!

**Review by fxdawnn — APPROVED** [review]

File: test/functorch/test_leaf_function.py [hunk 3]
```diff
@@ -2475,5 +2475,320 @@ def forward(self, x):
...
+
+        @leaf_function
+        def my_fn(x, tag, scale):
+            return (x * scale,)
+
+        @my_fn.register_fake
+        def my_fn_fake(x, tag, scale):
+            return (torch.empty_like(x),)
+
+        @my_fn.register_hook
```
Comment (ydwu4) [inline test/functorch/test_leaf_function.py:2545]: wait, i thought we'll no longer do register_hook, just do register_multi_grad_hook? I don't like register_hook because there are clearly multiple inputs to the leaf_function, which is better suited to do [register_multi_grad_hook](https://docs.pytorch.org/docs/stable/autograd.html#torch.autograd.graph.register_multi_grad_hook)

Comment (fxdawnn) [reply test/functorch/test_leaf_function.py:2545]: it's a mix of both register_grad and [register_multi_grad_hook](https://docs.pytorch.org/docs/stable/autograd.html#torch.autograd.graph.register_multi_grad_hook). Do you want to replace all the register_hook or leave the current ones as they are?

Comment (fxdawnn) [reply test/functorch/test_leaf_function.py:2545]: Even for a single tensor?

Comment (ydwu4) [reply test/functorch/test_leaf_function.py:2545]: yeah, we want register_multi_grad_hook even for a single tensor input.

**Comment by fxdawnn** [comment]
pivoted here -- https://github.com/pytorch/pytorch/pull/179609
